server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java

```diff
@@ -104,7 +153,7 @@ int numResponses() {
      * Returns the merged response. To be called once all responses have been added through {@link #add(SearchResponse)}
      * so that all responses are merged into a single one.
      */
-    SearchResponse getMergedResponse(Clusters clusters) {
+    synchronized SearchResponse getMergedResponse() {
```

Comment:
see comment above on the need for synchronized

Reply:
I reviewed the code more carefully and everything it calls (such as SearchProgressListener.ccsReduce). Without synchronization here everything it does and calls must be thread safe. I believe it is (as the AsyncTask.SearchProgressListener.ccsReduce call to MutableSearchThread.updatePartialResponse is `synchronized`.

So I believe the synchronization here can be removed (and on `add`) but it is a little more risky that any non-thread safe changes in the future will cause issues. I added code comments indicating that these two methods must remain thread-safe.

There are also tests already in SearchResponseTests that test thread safety. 

The one thing this "breaks" or interferes with is that that optimization to not do an incremental merge on the last SearchResponse sent to the `add` method. Say there are 2 responses still to come in and both call add at the same time. In that case, in the tests I'm getting non-deterministic results around how many incremental merges happened, since there is a (benign) race condition. I've adjusted the tests to only run asserts against the number of incremental merges if the test runs serially rather than concurrently. I ran the tests 100,000 times and they did not fail.

However, given the non-deterministic nature of the optimization now, requiring extra work in the tests and being initiall confusing to code readers/maintainers, I wonder if we should remove this optimization.